### PR TITLE
Stream Sharding amortize push payloads across the observed rate.

### PR DESCRIPTION
**Pull request by MasslessParticle**
This should account for large payloads that come infrequently

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -471,10 +471,11 @@ func (d *Distributor) divideEntriesBetweenShards(tenantID string, totalShards in
...
 	var (
-		streamLabels   = labelTemplate(stream.Labels)
-		streamPattern  = streamLabels.String()
-		derivedKeys    = make([]uint32, 0, totalShards)
-		derivedStreams = make([]streamTracker, 0, totalShards)
+		streamLabels    = labelTemplate(stream.Labels)
+		streamPattern   = streamLabels.String()
+		derivedKeys     = make([]uint32, 0, totalShards)
+		derivedStreams  = make([]streamTracker, 0, totalShards)
+		entriesPerShard = int(math.Ceil(float64(len(stream.Entries)) / float64(totalShards)))
```
Comment (DylanGuedes) [inline pkg/distributor/distributor.go:478]: by moving the `entriesPerShard` evaluation to this line you're doing it before the `totalShards <= 0` check, that is done at line 483. The check was added to guarantee against by 0 divisions.

That said, WDYT of moving the totalShards <= 0 to before the vars definition?

File: pkg/distributor/distributor_test.go [hunk 2]
```diff
@@ -806,70 +806,101 @@ func TestShardCountFor(t *testing.T) {
...
+			name:        "enough data to have two shards, stream size (36b) + ingested rate (0b) > 22b",
+			stream:      &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
+			rate:        0,  // in bytes
+			desiredRate: 22, // in bytes
+			pushSize:    36, // in bytes
+			wantShards:  2,
+			wantErr:     false,
 		},
 		{
 			name: "a lot of shards, stream size (1mb) + ingested rate (300mb) > 3mb",
```
Comment (DylanGuedes) [inline pkg/distributor/distributor_test.go:865]: could you fix the numbers on this test description? it should be:
```
push size (90b) + ingested rate (0b) > 22b
```

File: pkg/distributor/distributor_test.go [hunk 3]
```diff
@@ -806,70 +806,101 @@ func TestShardCountFor(t *testing.T) {
...
+			desiredRate: 22, // in bytes
+			pushSize:    90, // in bytes
+			wantShards:  5,
+			wantErr:     false,
+		},
+		{
+			name:        "take push rate into account. Only generate two shards even though this push is quite large",
+			stream:      &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
+			rate:        24,        // in bytes
+			pushRate:    1.0 / 6.0, // 6 pushes per second
```
Comment (DylanGuedes) [inline pkg/distributor/distributor_test.go:879]: this confused me a little, why 6 pushes per second isn't `6.0 / 1.0` instead? it sounds like the more I push, the lower the push rate; I'd expect the opposite

File: pkg/distributor/distributor_test.go [hunk 4]
```diff
@@ -806,70 +806,101 @@ func TestShardCountFor(t *testing.T) {
...
+			stream:      &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
+			rate:        24, // in bytes
+			pushRate:    0,
+			desiredRate: 40,  // in bytes
+			pushSize:    200, // in bytes
+			wantShards:  6,
+			wantErr:     false,
+		},
+		{
+			name:        "If the push rate is greater than 1, use the payload size",
```
Comment (DylanGuedes) [inline pkg/distributor/distributor_test.go:896]: this part confused me a little, why rate=1 has the same effect of rate=0?

File: pkg/distributor/ratestore_test.go [hunk 5]
```diff
@@ -84,32 +78,26 @@ func TestRateStore(t *testing.T) {
...
-		require.Eventually(t, func() bool { // There will be data
-			return tc.rateStore.RateFor("tenant 1", 0) != 0 &&
-				tc.rateStore.RateFor("tenant 2", 0) != 0
-		}, time.Second, time.Millisecond)
+		require.NoError(t, tc.rateStore.instrumentedUpdateAllRates(context.Background()))
 
-		require.Equal(t, int64(35), tc.rateStore.RateFor("tenant 1", 0))
-		require.Equal(t, int64(35), tc.rateStore.RateFor("tenant 2", 0))
+		requireRatesAndPushesEqual(t, 35, 10, tc.rateStore, "tenant 1", 0)
+		requireRatesAndPushesEqual(t, 35, 10, tc.rateStore, "tenant 1", 0)
```
Comment (DylanGuedes) [inline pkg/distributor/ratestore_test.go:97]: tenant 2 also expects 35 as the highest rate
```suggestion
		requireRatesAndPushesEqual(t, 35, 10, tc.rateStore, "tenant 2", 0)
```

File: pkg/distributor/ratestore_test.go [hunk 6]
```diff
@@ -173,17 +154,81 @@ func TestRateStore(t *testing.T) {
...
+
+		tc.clientPool.clients = map[string]client.PoolClient{
+			"ingester0": newRateClient([]*logproto.StreamRate{
+				{Tenant: "tenant 1", StreamHash: 1, StreamHashNoShard: 0, Rate: 50, Pushes: 50},
+			}, 1),
+		}
+
+		require.NoError(t, tc.rateStore.instrumentedUpdateAllRates(context.Background()))
+		afterNewRate, afterNewPushRate := tc.rateStore.RateFor("tenant 1", 0)
+		require.EqualValues(t, weightedMovingAverage(50, 25), afterNewRate)
```
Comment (DylanGuedes) [inline pkg/distributor/ratestore_test.go:198]: this is pretty cool, gj

**Review by DylanGuedes — COMMENTED** [review]
looks nice, a few questions

File: pkg/distributor/distributor_test.go (hunk 4 above)
Comment (MasslessParticle) [reply pkg/distributor/distributor_test.go:900]: I've added a comment there to try and explain. lmk if it helps clear things up

File: pkg/distributor/distributor_test.go (hunk 3 above)
Comment (MasslessParticle) [reply pkg/distributor/distributor_test.go]: The comment is backward. This represents 1 push every 6 seconds. Fixed

**Review by owen-d — APPROVED** [review]
a question but looks good, thanks!

File: pkg/distributor/ratestore.go [hunk 7]
```diff
@@ -206,6 +246,7 @@ func (s *rateStore) aggregateByShard(ctx context.Context, streamRates map[string
 		for _, streamRate := range tenant {
 			rate := rates[tID][streamRate.StreamHashNoShard]
 			rate.rate += streamRate.Rate
+			rate.pushes = math.Max(float64(streamRate.Pushes), rate.pushes)
```
[the remark this answers is not in the document]
Comment (MasslessParticle) [reply pkg/distributor/ratestore.go:249]: They're doing a couple of subtly different things: `ratesPerStream` deduplicates across replicas taking the max pushes/rates. `aggregateByShard` aggregates across the shards and adds the rates (to reconstruct the rate coming from the un-sharded stream) and takes the max pushes across shards (with the assumption that shards with higher push-counts more closely resemble the original stream's push rate)

**Pull request merged**